src/Common/Auth/OpenIDConnect/Entities/UserEntity.php

```diff
@@ -31,6 +32,22 @@ public function getClaims()
         $claimsType = (!empty($_REQUEST['grant_type']) && ($_REQUEST['grant_type'] === 'client_credentials')) ? 'client' : 'oidc';
         if ($claimsType === 'oidc') {
             $uuidToUser = new UuidUserAccount($this->identifier);
+            $fhirUser = '';
+            $userRole = $uuidToUser->getUserRole() == 'users';
+            if ($userRole == 'users') {
+                // TODO: adunsulag check with brady/sjpadget on whether this should be Practioner or Person, it has to
+                // be one of those resource types and you have to be able to retrieve it via a FHIR endpoint but I'm not
+                // sure a site admin is classified as a 'practioner'.
+                // TODO: adunsulag we should see if there is a better way like FHIRRouteResolver for a given resource endpoint...
+                $fhirUser = $GLOBALS['site_addr_oath'] . $GLOBALS['web_root'] . '/apis/' . $_SESSION['site_id'] . "/fhir/Practitioner/" . $this->identifier;
+            } else if ($userRole == 'patients') {
```

Comment:
@sjpadgett and @adunsulag ,
I think I'm getting confused on the userRole stuff (which also caused me to possibly go in the wrong direction on some changes i made in my last commit regarding Jerry's scope work). At this point, dispatch.php will send users to the standard fhir api route (fhir) and send patients to the patient-centric api route (fhirportal). When I created the patient-specific auth/api/fhir stuff in the past I created separate routes for patients to ensure that patients could never see other patient data (otherwise it was very difficult to guarantee this in the standard fhir and api routes). Is this a feasible strategy? And is my concept of userRole the same that is being used in the scopes?

Reply:
@bradymiller I've got another project I have to take care of this morning but this afternoon (EST time) I will see if I can push out a simple PR that shows an alternative approach to the routes being separate.

The user role here is talking about the role of the logged in user whether the user is a patient or comes from the administrator role.

If we wanted to keep things centralized for the FHIR api / standard api instead of having duplicate patient routes, we would key off the scope context parameter.  If a scope has a 'patient/' prefix we would force every route to have the current patient ID to prevent user data from leaking.  You have to do this anyways with the patient portal routes right?  So if I'm understanding the code and what I discussed with Jerry we would just centralize this.  I think it will make more sense at if I throw some kind of example PR out and then you can look at it from a security perspective and say yes that will work fine, or no that is going to cause problems.

Reply:
hi @adunsulag , That is likely a feasible strategy for the fhir route, but prob not for the standard api route (since those routes/code are much more complex), since fhir route is much more standard. Regarding security, I generally try to incorporate several layers to avoid systemic critical vulnerabilities, and it definitely makes me nervous that the only wall here would be essentially one piece of code (ie. that is broken in future, then it all falls down). The separate route allows me to sleep at night knowing I won't get roped into working on a critical vulnerability for several days (usually when I least want to work on one :) ), but my instinct tells me I can maybe convert the patient fhir route over to standard  fhir without messing up my future sleep. I'll put this in my queue to work on for 6.1 (no need for you to spend your time on this unless you are interested to derive a solution, of course).

Reply:
@bradymiller defense in depth is a good strategy, the flip side is complexity can also lead to security vulnerabilities especially if code is being duplicated in multiple places.  I'm fine with working on the other stuff, there's plenty to do.

Reply:
I still think this is a bad idea but promised myself I wasn't going to get too obstinate about it. If you feel strongly @bradymiller then that's the way it should be. (Although, my nightmares are generally stuff like falling off a cliff:))

I still need to refine the scope delegations by role to make things clearer. This includes adding the trusted users authenticated role to have a clear departure between client_role and user_role for scope validation purposes.

Reply:
@sjpadgett , fyi, couldn't help myself and looking into this now :) (noted a potential security issue already in the patient route that only has 3 endpoints so far; looking into that and also into how can force/bootstrap patient uuid in the requests and then see if can leverage that if can leverage that in the standard fhir route)

Reply:
I'm thinking for v6 release, we go a head and lock it down to just confidential client and OpenEMR sign in. Leave the patient role for v6.1. @bradymiller  and @adunsulag thoughts.

Reply:
Confidential client with support for Launch-EHR I think should be the target for v6.  Patient login & standalone patient launch apps should be pushed to v6.1 in my opinion.  That's what I've set in the SMART well-known capabilities support (we currently are broadcasting that we support the **user-permissions** capability but not **patient-permissions** right now).

Reply:
Okay then i'll leave with you to change the discovery, patch out the via Patient button in log in dialog and force default client registration to private app type. Unless you want me to do after you finish this PR and @bradymiller agrees.

Reply:
Sounds good to me. Regarding current patient routes (turns out the current 3 patient endpoints are secure after completing my audit today), one option would be to just make the 2 globals that allow the patient routes to be EXPERIMENTAL (and if these are off, there is the no way for patient roles to be functional; and could ensure either one of these are turned on when offering the Patient button in log in authorize.php) along with updating docs. Then really no codebase differences between master/rel-600 so easier to bring things back to rel-600 (if wish to get in a patch) as figure best way to support patient role in master.

Reply:
regarding above, just let me know, and can include it in PR i am currently working on

Reply:
Okay Brady, i'll leave with you guys.
lol, kind of feels like when one of the kids go off to college ie hard to let go!:) Still you know they'll be back when they need money... Same here if need help.

Reply:
@bradymiller I'm fine with turning it off if you want to do things in your PR.  Unfortunately, I had a customer emergency crop up that is consuming more of today than I thought.  I probably won't be able to revisit this until tomorrow unfortunately.

Reply:
no prob @adunsulag . Brought the related changes into the PR here (basically downplays use of the patient role/route/endpoint to experimental but still allows use if user/developer really wants to):
https://github.com/openemr/openemr/pull/4132